packages/connection/src/class-nonce-handler.php

```diff
@@ -0,0 +1,197 @@
+<?php
+/**
+ * The nonce handler.
+ *
+ * @package automattic/jetpack-connection
+ */
+
+namespace Automattic\Jetpack\Connection;
+
+/**
+ * The nonce handler.
+ */
+class Nonce_Handler {
+
+	/**
+	 * How many nonces should be removed during each run of the runtime cleanup.
+	 * Can be modified using the filter `jetpack_connection_nonce_cleanup_runtime_limit`.
+	 */
+	const CLEANUP_RUNTIME_LIMIT = 10;
+
+	/**
+	 * How many nonces should be removed per batch during the `clean_all()` run.
+	 */
+	const CLEAN_ALL_LIMIT_PER_BATCH = 1000;
+
+	/**
+	 * Nonce lifetime in seconds.
+	 */
+	const LIFETIME = HOUR_IN_SECONDS;
+
+	/**
+	 * The nonces used during the request are stored here to keep them valid.
+	 *
+	 * @var array
+	 */
+	private static $nonces_used_this_request = array();
+
+	/**
+	 * Adds a used nonce to a list of known nonces.
+	 *
+	 * @param int    $timestamp the current request timestamp.
+	 * @param string $nonce the nonce value.
+	 * @param bool   $run_cleanup Whether to run the `cleanup_runtime()`.
+	 *
+	 * @return bool whether the nonce is unique or not.
+	 */
+	public static function add( $timestamp, $nonce, $run_cleanup = true ) {
+		global $wpdb;
+
+		if ( isset( static::$nonces_used_this_request[ "$timestamp:$nonce" ] ) ) {
+			return static::$nonces_used_this_request[ "$timestamp:$nonce" ];
+		}
+
+		// This should always have gone through Jetpack_Signature::sign_request() first to check $timestamp an $nonce.
```

Comment:
Wow, that type has been there since the beginning of Jetpack: https://github.com/Automattic/jetpack/commit/99e10babcb00e4d5f9a8d9d91584f71626fceb2f
Literally the first commit 😀